feat(stdlib): Allow arbitrary length sha256 hashes

bergkvist opened the pull request:
`std::crypto::hashes::sha256` only exposed 2 procedures before this PR - `sha256::hash_1to1` and `sha256::hash_2to1`. These apply static padding to the input, which means input has to be exactly 256 bit or 512 bit. Any other input length could not be hashed using the existing interface.

prepare_message_schedule_and_consume had previously hardcoded values that meant it only worked for `i = 0`. This PR generalizes it to work for any loop iteration (not just i = 0), and exposes it.

```
use.std::crypto::hashes::sha256
begin
  # Sha256 padding for bitcoin block header
  push.0x00000280.0x00000000.0x00000000.0x00000000
  push.0x00000000.0x00000000.0x00000000.0x00000000
  push.0x00000000.0x00000000.0x00000000.0x80000000
  
  # Bitcoin block header (640 bits)
  push.0x008d0d90.0xffff001d.0x023a124a.0x328ea7e8
  push.0x5dc3de1b.0x7084ab7e.0x07fbb06e.0xe4895ff5
  push.0x4ebfa924.0xfaf22367.0x3174efe8.0x00000000
  push.0x94eca277.0x23a42848.0x8931a006.0x25b55d62
  push.0xea642fba.0x6b29ee13.0x305e4ef4.0x01000000
  
  # Sha256 initial state
  push.0x5be0cd19.0x1f83d9ab.0x9b05688c.0x510e527f
  push.0xa54ff53a.0x3c6ef372.0xbb67ae85.0x6a09e667
  
  # Compute hash of bitcoin header
  exec.sha256::prepare_message_schedule_and_consume # ( i = 0 )
  exec.sha256::prepare_message_schedule_and_consume # ( i = 1 )
  
  # Hashing again gives us the BTC block hash
  exec.sha256::hash_1to1
  
  push.0xb73860e7 assert_eq
  push.0xceeedf95 assert_eq
  push.0xd07e8bdd assert_eq
  push.0x477fb868 assert_eq
  push.0x1fb22dfb assert_eq
  push.0xe2d3c3a7 assert_eq
  push.0xbd48355c assert_eq
  push.0x00000000 assert_eq
end
```

After your input has been padded, you simply execute `sha256::prepare_message_schedule_and_consume` once for every 512-bit block of padded input.

bobbinth reviewed (commented):
This looks great! Thank you!

One question: instead of exposing `prepare_message_schedule_and_consume` (which seems to be a rather low-level procedure), would it be better to expose a higher-level procedures  - e.g., something like `update` and `finalize` or maybe just `hash_memory` (which would hash the data from memory)? Or would that be to big of a change?

bobbinth commented on stdlib/asm/crypto/hashes/sha256.masm:
```diff
@@ -232,11 +232,13 @@ end
-#! & https://github.com/itzmeanjan/merklize-sha/blob/8a2c006/include/sha2_256.hpp#L148-L187 ( loop body execution when i = 0 )
-proc.prepare_message_schedule_and_consume.2
+#! & https://github.com/itzmeanjan/merklize-sha/blob/8a2c006/include/sha2_256.hpp#L148-L187 ( loop body execution )
+export.prepare_message_schedule_and_consume.4
```
Should we change this back to `proc`?

bobbinth reviewed (approved):
This is awesome! Thank you! Ideally, would be great to add a test for this - but we can do it later.

bergkvist commented:
@bobbinth I've added a test for `sha256::hash_memory` as well now - and changed `export.prepare_message_schedule_and_consume` to `proc.prepare_message_schedule_and_consume`.

bobbinth commented:
> I've added a test for `sha256::hash_memory` as well now - and changed `export.prepare_message_schedule_and_consume` to `proc.prepare_message_schedule_and_consume`.

Thank you for putting these together!
